src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -0,0 +1,246 @@
+            {
+                if (_disposeTokenSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                if (unit is RemovalWork removalWork)
+                {
+                    await ReportRemovalAsync(removalWork, cancellationToken).ConfigureAwait(false);
+                }
+
+                var project = solution.GetProject(unit.ProjectId);
+                if (project is null)
+                {
+                    _logger?.LogTrace("Project {projectId} is not in workspace", unit.ProjectId);
+                    continue;
+                }
+
+                await UpdateProjectAsync(project, solution, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                _logger?.LogError(ex, "Encountered exception while processing unit: {message}", ex.Message);
+            }
+        }
+
+        await _stream.AssumeNotNull().FlushAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    private void Workspace_WorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+    {
+        switch (e.Kind)
+        {
+            case WorkspaceChangeKind.SolutionChanged:
+            case WorkspaceChangeKind.SolutionReloaded:
+                foreach (var project in e.NewSolution.Projects)
+                {
+                    EnqueueUpdate(project);
+                }
+
+                break;
+
+            case WorkspaceChangeKind.SolutionAdded:
+                foreach (var project in e.NewSolution.Projects)
+                {
+                    EnqueueUpdate(project);
+                }
+
+                break;
+
+            case WorkspaceChangeKind.ProjectReloaded:
+                EnqueueUpdate(e.NewSolution.GetProject(e.ProjectId));
+                break;
+
+            case WorkspaceChangeKind.ProjectRemoved:
+                RemoveProject(e.OldSolution.GetProject(e.ProjectId.AssumeNotNull()).AssumeNotNull());
+                break;
+
+            case WorkspaceChangeKind.ProjectAdded:
+            case WorkspaceChangeKind.ProjectChanged:
+            case WorkspaceChangeKind.DocumentAdded:
+            case WorkspaceChangeKind.DocumentRemoved:
+            case WorkspaceChangeKind.DocumentReloaded:
+            case WorkspaceChangeKind.DocumentChanged:
+            case WorkspaceChangeKind.AdditionalDocumentAdded:
+            case WorkspaceChangeKind.AdditionalDocumentRemoved:
+            case WorkspaceChangeKind.AdditionalDocumentReloaded:
+            case WorkspaceChangeKind.AdditionalDocumentChanged:
+            case WorkspaceChangeKind.DocumentInfoChanged:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentAdded:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentRemoved:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentReloaded:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentChanged:
+                var projectId = e.ProjectId ?? e.DocumentId?.ProjectId;
```

Comment:
Was this just defensively written or are there cases where we might actually have a null ProjectId and a non-null DocumentId? I wouldn't think that was possible, but maybe?

Reply:
Looks like I added it without much fanfare in https://github.com/dotnet/razor/pull/8607, so its possibly overly defensive, but this was written when we were having to be a bit secretive so if there was a real scenario, we didn't write it down anywhere.